src/clj/rems/workflow/dynamic.clj

```diff
@@ -544,16 +547,24 @@
   [user-ids injections]
   (apply merge-with into (keep (partial invalid-user-error injections) user-ids)))
 
-(defn- validation-error
-  [application injections]
-  (when-let [errors ((:validate-form injections) (:id application))]
-    {:errors errors}))
+(defn- validation-error [application {:keys [get-form]}]
+  (let [form (get-form (:form/id application))
+        _ (assert form)
+        answers (:form-contents application)
+        legacy-form {:items (for [field (:items form)]
+                              (assoc field :value (get-in answers [:items (:id field)])))
+                     :licenses (for [license (:application/licenses application)]
+                                 {:id (:license/id license)
+                                  :approved (= "approved"
+                                               (get-in answers [:licenses (:license/id license)]))})}
+        result (form-validation/validate legacy-form)]
```

Comment:
hmm, it still might be better design to inject the whole validate-form function, but have this as its implementation

Reply:
The form validation is a pure function, so there is little need to inject it.

Reply:
I'll extract it to make workflows unaware of form details.